Enable computing a full image tree with quickshift

SebastienEske opened the pull request:
Signed-off-by: Sebastien ESKENAZI <s.eskenaz@pixelz.com>

## Description

This pull requests adds the option `full_search` to the function `quickshift` in order to produce a complete hierarchy of the image superpixels as in the initial paper.

[1] Quick shift and kernel methods for mode seeking,
           Vedaldi, A. and Soatto, S.
           European Conference on Computer Vision, 2008

## Checklist

- [x] Gallery example in `./doc/examples` (new features only)
- [ ] Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- [x] Unit tests
- [x] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

## For reviewers

<!-- Don't remove the checklist below. -->
- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [x] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.
- [ ] Consider backporting the PR with `@meeseeksdev backport to v0.14.x`

SebastienEske committed:
Enable computing a full image tree with quickshift

Signed-off-by: Sebastien ESKENAZI <s.eskenaz@pixelz.com>

pep8speaks commented:
Hello @SebastienEske! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`skimage/segmentation/_quickshift.py`](https://github.com/scikit-image/scikit-image/blob/f9a84e0c5a1118c9c8c8060f40c5c7626a60a3f2/skimage/segmentation/_quickshift.py):

> [Line 12:77](https://github.com/scikit-image/scikit-image/blob/f9a84e0c5a1118c9c8c8060f40c5c7626a60a3f2/skimage/segmentation/_quickshift.py#L12): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 40:75](https://github.com/scikit-image/scikit-image/blob/f9a84e0c5a1118c9c8c8060f40c5c7626a60a3f2/skimage/segmentation/_quickshift.py#L40): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 72:58](https://github.com/scikit-image/scikit-image/blob/f9a84e0c5a1118c9c8c8060f40c5c7626a60a3f2/skimage/segmentation/_quickshift.py#L72): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace

emmanuelle commented:
Thank you for your contribution @SebastienEske ! Before we dive into your code, could you please explain how to use the returned tree in an image processing workflow. Ideally you would provide an example script which could be transformed to a gallery example or part of an existing one.

SebastienEske committed:
Fix PEP issues and add example for quickshift full_tree option

Signed-off-by: Sebastien ESKENAZI <s.eskenaz@pixelz.com>

SebastienEske committed:
Improve quick_shift full_tree option

Fix bug in full_tree option of quickshift
Fix PEP compliance
Add unit test cases

Signed-off-by: Sebastien ESKENAZI <s.eskenaz@pixelz.com>

SebastienEske committed:
Add quickshift full_search/single tree feature

Signed-off-by: Sebastien ESKENAZI <s.eskenaz@pixelz.com>

SebastienEske committed:
Fix typo bug in quickshift test

Signed-off-by: Sebastien ESKENAZI <s.eskenaz@pixelz.com>

SebastienEske committed:
Fix bug in quickshift test for full_search option

Signed-off-by: Sebastien ESKENAZI <s.eskenaz@pixelz.com>

SebastienEske committed:
Add graylevel test case for quickshift with full_search enabled

And fix typos and build issues with missing blank lines.

Signed-off-by: Sebastien ESKENAZI <s.eskenaz@pixelz.com>

SebastienEske commented:
Hello @emmanuelle all is fixed and done I believe. Sorry for the multiple commits. Do you want me to squash them all in one?

emmanuelle commented:
@SebastienEske don't worry about the multiple commits we can squash them at merge time :-). Now I just need to review again ;-)

emmanuelle commented on doc/examples/segmentation/plot_quickshift_full_tree.py:
```diff
@@ -0,0 +1,161 @@
+One of the benefits of quickshift is that it actually computes a
+hierarchical segmentation on multiple scales simultaneously.
+
+Quickshift has two main parameters: ``sigma`` controls the scale of the local
```
It is called `kernel_size` in the `quickshift` function

SebastienEske replied:
I am ok to change it but just so you know this is a copy paste from [the generic segmentation example](https://github.com/scikit-image/scikit-image/blob/d8f23c64c45427962ec5e7f975140de2d726547d/doc/examples/segmentation/plot_segmentations.py#L37). Do you want me to update it as well?

emmanuelle commented:
Thank you for the gallery example ! It is very dense and after a quick read through I'm having a hard time understanding what each figure shows. Could you please add some explanations about what each plot shows? What is the multiscale and multilevel segmentation? How can one use this type of segmentation (wih machine learning?)?

emmanuelle commented:
I'll try to dive deeper into the example later but if you could help me with additional comments it'd be great.

SebastienEske committed:
Add explanations and comment in the example for quickshift full_tree

Signed-off-by: Sebastien ESKENAZI <s.eskenaz@pixelz.com>

SebastienEske commented:
I've added the comments both in the code and on the figures. Basically, scale/distance segmentation is the usual way to make superpixels with quickshift while "level" segmentation simply shows the hierarchy levels in the tree returned by quickshift (from bottom to top in the example).

Borda commented on doc/examples/segmentation/plot_quickshift_full_tree.py:
```diff
@@ -0,0 +1,171 @@
+    a.set_axis_off()
+
+plt.tight_layout()
+plt.show()
```
pls, add the output figure to this image :)

SebastienEske replied:
Hello, thanks for the review :-)
I am not sure that I understand this properly. You want me to save the output as a png (or some other file format) and display it?

sciunto replied:
I think borda asked for posting the figure on github.

SebastienEske replied:
Ah! Ok. Here are all the figures. Left with full_search=True, and right with full_search=False. Output from quickshift
![ouput](https://user-images.githubusercontent.com/53205950/66184759-67635200-e6a7-11e9-9411-2e49983e1933.png)
Tree hierarchy segmentation (bottom to top)
![level](https://user-images.githubusercontent.com/53205950/66184766-69c5ac00-e6a7-11e9-98fe-82886126a411.png)
Usual distance based superpixel segmentation
![scale](https://user-images.githubusercontent.com/53205950/66184772-6b8f6f80-e6a7-11e9-82f8-bfb216a62ec7.png)

emmanuelle commented:
So, how would one typically use the different scales in a segmentation or machine learning problem? Sorry to ask naive questions, with just one array of superpixels I understand how one would for example compute features of the superpixels and then run a classification estimator on them, but how about the more complex information returned by the full image tree? If you feel like it would be easier to explain on another kind of image (medical image, etc.) we can restart here the discussion on having more example images ;-).

SebastienEske commented:
Well, quickshift already returns a tree, so I think the question is about what use is a full image tree compared to the current tree that is returned. Simply put, if a user wants a hierarchy of the image content (either for color anlaysis, image compression, or anything else), having the image split into several trees (current situation, "orphaned roots" mentioned in the TODO) is not much use as there is no relative hierarchy between those trees. Having a full image tree solves this problem.

SebastienEske commented:
Another advantage of the full tree is that it retrieves all the images scales (a notorious issue with superpixels). The user can then compute the segmentation for a set of scales and choose the one he/she wants.

Here are the results on an HD image (1920*1080) [from pixabay](https://pixabay.com/photos/camp-camping-nature-tent-freedom-4522970/) so it is free of rights.

Algorithm output, even with full search it is not just one region because of the `max_dim` parameter
![output](https://user-images.githubusercontent.com/53205950/66282920-99fd8c80-e8eb-11e9-8a74-d10245ef1abc.png)
Usual superpixel segmentation up to the maximum image scale of 2048 for which we actually have only one region. Notice the obvious limitations of not using `full_search`: it only goes as far as scale 64.
![superpixels](https://user-images.githubusercontent.com/53205950/66282918-9964f600-e8eb-11e9-9055-5bd81305293e.png)
Level/hierarchy segmentation. The results seem a bit more meaningful with `full_search` again although I would not bet on it:
![levelseg](https://user-images.githubusercontent.com/53205950/66282919-9964f600-e8eb-11e9-808b-6c3a1ba91031.png)

If you want we can use this image instead of the astronaut image to make the example more obvious.

SebastienEske commented:
Hello,
Any updates?

lagru commented:
Hey @SebastienEske, sorry for the silence and thank you for your continued efforts! Unfortunately I don't have the bandwidth right now to do a proper review on this. But we have not forgotten you! :wink: 

In the meantime I'm pinging @scikit-image/core to make sure this doesn't get buried.

SebastienEske commented:
Ok, thanks!

matthew-brett commented:
We have changed the base branch for this PR from the `main` branch to the `pre-megamove-integration` branch.

## Summary - what this means for your PR

You can continue to work on your PR, now based on `pre-megamove-integration`.  You should update your branch to the current state of `pre-megamove-integration` (this is what the Github "Update branch" button does).   Please **do not merge the scikit-image `main` branch into your PR branch**.

Once we've merged your PR into the `pre-megamove-integration` branch, we'll update your changes for Megamove, and merge into the `main` branch.

In addition - our automated script detected that your branch had merge conflicts with the pre-Megamove branch. Please make sure to merge in the `pre-megamove-integration` branch, and resolve the merge conflicts in the way that you want.

Read on for the Megamove rationale, and more detailed instructions.

<details>

## Rationale

We've changed the base branch to make it easier and smoother to continue to work on your PR, while we work through a rather large change in the directory structure of the repository - a change that we have called *Megamove*.

See [the Megamove explanation issue](https://github.com/scikit-image/scikit-image/issues/8232) for more detail.

Changing the base branch means you do not see merge conflicts related to the *Megamove* change, and can continue to work with the repository in its original state, without worrying about porting those changes across to the new post-Megamove structure.

As you'll see in the linked issue, you can continue to work on your PR as before, and you can continue to use the Github GUI to "Update branch" - this updates from the new `pre-megamove-integration` branch.   However, there is one caveat: please, do not try and merge the `scikit-image` `main` branch into this one, but instead, keep track of the `pre-megamove-integration` branch.  The issue explains more.

To emphasize again, **do not try and merge the scikit-image `main` branch**.  You will get Megamove related merge conflicts that are rather tricky to resolve correctly.

When we've approved your PR, we'll merge into the `pre-megamove-integration` branch, and then process it so we can merge it into the `main` branch.   We'll let you know when we've merged your PR changes into `main`.

</details>

Any questions, please reply here.
